### PR TITLE
C-API output tensors reset

### PR DESCRIPTION
**Pull request by atobiszei**
### 🛠 Summary
Adding safeguard to C-API so that when setting output buffers is used we restore the original output buffers after the inference. Otherwise you would have to use output tensors always for any particular model. ov::InferRequest would store the previously set output buffer, and inference would trigger overwrite.

Ticket:CVS-141788, CVS-152665
### 🧪 Checklist

- [x] Unit tests added.
- [ ] The documentation updated.
- [x] Change follows security best practices.
``

### CODE REVIEW

File: src/test/openvino_remote_tensors_tests.cpp [hunk 1]
```diff
@@ -1269,15 +1275,203 @@ TEST_F(CAPINonCopy, SyncWithCallbackDummy) {
...
+    std::vector<int> expectedShape{1, 10};
+    for (size_t i = 0; i < DUMMY_MODEL_SHAPE.size(); ++i) {
+        EXPECT_EQ(expectedShape[i], shape[i]) << "Different at:" << i << " place.";
+    }
+    const float* outputData = reinterpret_cast<const float*>(voutputData);
+    for (int i = 0; i < expectedShape[1]; ++i) {
+        EXPECT_NEAR(expectedValue, outputData[i], tolerance) << "Different at:" << i << " place.";
+    }
+}
+TEST_F(CAPINonCopy, SyncWithCallbackDummyCheckResetOutputGPU) {
```
Comment (atobiszei) [inline src/test/openvino_remote_tensors_tests.cpp:1327]:
```suggestion
TEST_F(CAPINonCopy, OpenCL_SyncWithCallbackDummyCheckResetOutputGPU) {
```

File: src/test/openvino_remote_tensors_tests.cpp [hunk 2]
```diff
@@ -1269,15 +1275,203 @@ TEST_F(CAPINonCopy, SyncWithCallbackDummy) {
...
+    ASSERT_CAPI_STATUS_NULL(OVMS_InferenceRequestOutputRemoveData(request, DUMMY_MODEL_OUTPUT_NAME));
+    ASSERT_CAPI_STATUS_NULL(OVMS_InferenceRequestRemoveOutput(request, DUMMY_MODEL_OUTPUT_NAME));
+    ASSERT_CAPI_STATUS_NULL(OVMS_Inference(cserver, request, &response));
+    checkDummyResponse(response, INITIAL_VALUE + 42 + 1, FLOAT_TOLLERANCE);
+    // intentional check for original output buffer if they were not overridden
+    for (size_t i = 0; i < out1.size(); ++i) {
+        EXPECT_NEAR(INITIAL_VALUE + 1, out1[i], FLOAT_TOLLERANCE) << "Different at:" << i << " place.";
+    }
+    OVMS_InferenceResponseDelete(response);
+    // TODO tests for remove RequestOutputRemoveData
```
Comment (atobiszei) [inline src/test/openvino_remote_tensors_tests.cpp:1426]:
```suggestion
```

File: src/modelinstance.cpp [hunk 3]
```diff
@@ -1397,6 +1428,26 @@ Status ModelInstance::infer(const RequestType* requestProto,
     // handleCallback(requestProto, responseProto); to be enabled when callbacks are implemented in network API's
     return status;
 }
+void ModelInstance::checkForOutputTensorResetAbility() {
```
Comment (rasapala) [inline src/modelinstance.cpp:1431]: Add description what is the condition.

**Review by rasapala — APPROVED** [review]

File: src/BUILD [hunk 4]
```diff
@@ -1578,7 +1578,7 @@ cc_library(
     srcs = ["test/openvino_remote_tensors_tests.cpp"],
     data = [
         "test/c_api/config_standard_dummy.json",
```
Comment (mzegla) [inline src/BUILD:1580]: How about standard dummy config?

File: src/BUILD [hunk 5]
```diff
@@ -1627,7 +1627,7 @@ cc_library(
     srcs = ["test/openvino_tests.cpp"],
     data = [
         "test/c_api/config_standard_dummy.json",
```
Comment (mzegla) [inline src/BUILD:1629]: How about standard dummy config?

File: src/capi_frontend/capi.cpp [hunk 6]
```diff
@@ -693,6 +693,21 @@ DLL_PUBLIC OVMS_Status* OVMS_InferenceRequestRemoveInput(OVMS_InferenceRequest*
...
+    }
+    if (outputName == nullptr) {
+        return reinterpret_cast<OVMS_Status*>(new Status(StatusCode::NONEXISTENT_PTR, "output name"));
+    }
+    InferenceRequest* request = reinterpret_cast<InferenceRequest*>(req);
+    auto status = request->removeOutput(outputName);
+    if (!status.ok()) {
+        return reinterpret_cast<OVMS_Status*>(new Status(std::move(status)));
+    }
+    return nullptr;
```
Comment (mzegla) [inline src/capi_frontend/capi.cpp:708]: Is `nullptr` interpreted as OK status?

File: src/capi_frontend/inferencerequest.cpp [hunk 7]
```diff
@@ -60,6 +60,13 @@ Status InferenceRequest::setOutputBuffer(const char* name, const void* addr, siz
     }
     return it->second.setBuffer(addr, byteSize, bufferType, deviceId);
 }
+Status InferenceRequest::removeOutputBuffer(const char* name) {
+    auto it = outputs.find(name);
```
Comment (mzegla) [inline src/capi_frontend/inferencerequest.cpp:64]: Does it make sense to check `name` pointer validity first?

File: src/capi_frontend/inferencerequest.cpp [hunk 8]
```diff
@@ -99,6 +106,13 @@ Status InferenceRequest::removeInput(const char* name) {
     }
     return StatusCode::NONEXISTENT_TENSOR_FOR_REMOVAL;
 }
+Status InferenceRequest::removeOutput(const char* name) {
```
Comment (mzegla) [inline src/capi_frontend/inferencerequest.cpp:109]: same as above

File: src/modelinstance.cpp [hunk 9]
```diff
@@ -1397,6 +1428,26 @@ Status ModelInstance::infer(const RequestType* requestProto,
...
+    StreamIdGuard guard(getInferRequestsQueue());
+    auto request = guard.getInferRequest();
+    bool allOutputsSupported = true;
+    for (auto [name, _] : getOutputsInfo()) {
+        try {
+            ov::Tensor tensor = request.get_tensor(name);
```
Comment (mzegla) [inline src/modelinstance.cpp:1437]: Is it possible that we capture input name here? If 	model has an input called `name`?

File: src/modelinstance.cpp (hunk 3 above)
Comment (atobiszei) [inline src/modelinstance.cpp:1431]:
```suggestion
void ModelInstance::checkForOutputTensorResetAbility() {
    // for reseting output we need copy constructor of compiled model infer request ov::Tensor to be implemented.
```

File: src/modelinstance.cpp (hunk 9 above)
Comment (atobiszei) [reply src/modelinstance.cpp:1437]: Output name is logged in L1442/1439

File: src/capi_frontend/capi.cpp (hunk 6 above)
Comment (atobiszei) [reply src/capi_frontend/capi.cpp:708]: yes

File: src/capi_frontend/inferencerequest.cpp (hunk 7 above)
Comment (atobiszei) [reply src/capi_frontend/inferencerequest.cpp:64]: It's checked at capi.cpp

File: src/capi_frontend/inferencerequest.cpp (hunk 8 above)
Comment (atobiszei) [reply src/capi_frontend/inferencerequest.cpp:109]: It's checked at capi.cpp

**Review by mzegla — APPROVED** [review]

**Pull request merged**